# EAU

This is a repository for data analysis code associated with **Single-cell transcriptomic analysis of retinal immune regulation and blood-retinal barrier function during experimental autoimmune uveitis**.

Most code follows standard pipelines found in documentation for the following packages:
- Seurat (https://satijalab.org/seurat/) 
- SoupX (https://github.com/constantAmateur/SoupX)
- DoubletFinder (https://github.com/chris-mcginnis-ucsf/DoubletFinder)
- CellPhoneDB (https://github.com/ventolab/CellphoneDB)
- fgsea (https://bioconductor.org/packages/release/bioc/html/fgsea.html)

## Workflow
![workflow](./Misc/workflow.png)

## Methods (from paper)
### Preparation of retinas for single-cell RNA-sequencing
After imaging at day 21 post-immunisation, mice were euthanised by cervical dislocation and eyes were enucleated. The retinas were carefully dissected out and placed into Hanks’ Balanced Salt Solution (HBSS) containing 10 mM HEPES on ice for transport. After all retinas were collected, they were dissociated using the Worthington Papain Dissociation system. Retinas were placed in a solution of 20 U/ml papain, 0.005% DNase I with 1 mM L-cysteine and 5 mM EDTA in Earle’s Balanced Salt Solution (EBSS), for 10 minutes at 37°C with frequent, gentle agitation. Samples were then diluted by addition of 500 μl of EBSS to inactivate the papain and centrifuged at 300 x g for 5 min at room temperature. Pellets were resuspended in 525 μl of a solution containing 1 mg/ml ovomucoid and BSA and 100 U/ml DNase I in EBSS. The resulting suspension was carefully layered over 500 μl ovomucoid/BSA solution and centrifuged at 70 x g for 6 min. Supernatant was discarded and cells were resuspended in a solution of PBS containing 0.04% BSA.

### Single-cell RNA-seq library generation and processing
Library generation from dissociated retinas and initial data processing was performed by the Oxford Genomics Centre. scRNAseq transcriptome processing was performed using the Chromium 10X system involving GEM generation, post GEM-generation clean-up, cDNA amplification and DNA quantification. Chromium Single Cell Reagent Kits solution (10X SC RNA CITE-TSC, 10X SC RNA 5pr v2.0 Chemistry) was used to deliver a scalable microfluidic platform for digital scRNA-seq by profiling 500-10,000 individual cells per sample. A pool of ~3,500,000 10x Barcodes were sampled separately to index each cell’s transcriptome. Libraries were generated and sequenced from the cDNAs and 10x Barcodes were used to associate individual reads back to the individual partitions. The library was sequenced using the Illumina NovaSeq platform.
For initial data processing, Illumina’s bcl2fastq and cellranger mkfastq demultiplexes were used to convert the raw base call (BCL) files generated by Illumina sequencers into FASTQ files. Cellranger was used to perform alignment, filtering, barcode counting, and UMI counting from FASTQ files. This generated feature-barcode matrices for each sample, which was used for downstream analyses.

### Analysis of single-cell RNA-seq data
The SoupX package <sup>42</sup> was used to correct raw feature-barcode matrices for ambient RNA contamination. Corrected matrices were then analysed using the Seurat package 43. Commonly used QC metrics, such as UMI count, number of features, percentage mitochondrial RNA and percentage ribosomal RNA were used to filter out low quality cells. Samples were then normalised using the SCTransform function in Seurat, and integrated to remove batch effects and generate a single feature-barcode matrix.
The integrated matrix was then passed through standard dimensionality reduction and clustering pipelines in Seurat. Briefly, Principal Component Analysis (PCA) was used to determine dataset dimensionality, followed by shared nearest-neighbour graph construction and dimensionality reduction with the Uniform Manifold Approximation Projection (UMAP) method. Visualisations were performed using in-built functions in Seurat. Differential expression was performed by pseudobulking each identified cell type by sample and analysing with the DESeq2 package 44. Ligand-receptor interaction analysis was performed with CellPhoneDB package in Python using the statistical analysis method 45. DESeq2 and CellPhoneDB heatmaps were produced using the pheatmap package in R.
